contracts/staking/ServiceStakingBase.sol

```diff
@@ -0,0 +1,384 @@
+    event Checkpoint(uint256 availableRewards, uint256 numServices);
+    event ServiceUnstaked(uint256 indexed serviceId, address indexed owner, uint256 reward, uint256 tsStart);
+    event Deposit(address indexed sender, uint256 amount, uint256 balance, uint256 availableRewards);
+    event Withdraw(address indexed to, uint256 amount);
+
+    // Rewards per second
+    uint256 public immutable rewardsPerSecond;
+    // Minimum deposit value for staking
+    uint256 public immutable minStakingDeposit;
+    // Liveness ratio in the format of 1e18
+    uint256 public immutable livenessRatio;
+    // ServiceRegistry contract address
+    address public immutable serviceRegistry;
+
+    // Token / ETH balance
+    uint256 public balance;
+    // Token / ETH available rewards
+    uint256 public availableRewards;
+    // Timestamp of the last checkpoint
+    uint256 public tsCheckpoint;
+    // Minimum token / ETH balance, will be sent along with unstaked reward when going below that balance value
+    uint256 public minBalance;
+    // Mapping of serviceId => staking service info
+    mapping (uint256 => ServiceInfo) public mapServiceInfo;
+    // Set of currently staking serviceIds
+    uint256[] public setServiceIds;
+
+    /// @dev ServiceStakingBase constructor.
+    /// @param _rewardsPerSecond Staking rewards per second (in single digits).
+    /// @param _minStakingDeposit Minimum staking deposit for a service to be eligible to stake.
+    /// @param _livenessRatio Staking ratio: number of seconds per nonce (in 18 digits).
+    /// @param _serviceRegistry ServiceRegistry contract address.
+    constructor(uint256 _rewardsPerSecond, uint256 _minStakingDeposit, uint256 _livenessRatio, address _serviceRegistry) {
+        // Initial checks
+        if (_rewardsPerSecond == 0 || _minStakingDeposit == 0 || _livenessRatio == 0) {
+            revert ZeroValue();
+        }
+        if (_serviceRegistry == address(0)) {
+            revert ZeroAddress();
+        }
+
+        rewardsPerSecond = _rewardsPerSecond;
+        minStakingDeposit = _minStakingDeposit;
+        livenessRatio = _livenessRatio;
+        serviceRegistry = _serviceRegistry;
+    }
+
+    /// @dev Checks token / ETH staking deposit.
+    /// @param stakingDeposit Staking deposit.
+    function _checkTokenStakingDeposit(uint256, uint256 stakingDeposit) internal view virtual {
+        // The staking deposit derived from a security deposit value must be greater or equal to the minimum defined one
+        if (stakingDeposit < minStakingDeposit) {
+            revert LowerThan(stakingDeposit, minStakingDeposit);
+        }
+    }
+
+    /// @dev Withdraws the reward amount to a service owner.
+    /// @param to Address to.
+    /// @param amount Amount to withdraw.
+    function _withdraw(address to, uint256 amount) internal virtual {}
+
+    /// @dev Stakes the service.
+    /// @param serviceId Service Id.
+    function stake(uint256 serviceId) external {
```

Comment:
revert() if availableRewards is zero?

Reply:
I disagree, services can stake before rewards appear such that they are already in the staking accounting

Reply:
Nobody promised any rewards in the first place.

Reply:
IMO: It would seem more logical to me that staking would be allowed when availableRewards > 0. Rewards are paid from these according to the scheme in the current contract. While availableRewards >0, you can still add stackers. When all funds are exhausted during the next rewards distribution cycle, the contract goes on "pause" (nobody can stake). And so on until the contract is replenished via donate ETH/Token. Then the cycle resumes.